src/compiler/checker.ts

```diff
@@ -14715,7 +14715,7 @@ namespace ts {
             // we give type 'any[]' to 'x' instead of using the type determined by control flow analysis such that operations
             // on empty arrays are possible without implicit any errors and new element types can be inferred without
             // type mismatch errors.
-            const resultType = getObjectFlags(evolvedType) & ObjectFlags.EvolvingArray && isEvolvingArrayOperationTarget(reference) ? anyArrayType : finalizeEvolvingArrayType(evolvedType);
+            const resultType = getObjectFlags(evolvedType) & ObjectFlags.EvolvingArray && isEvolvingArrayOperationTarget(reference) ? autoArrayType : finalizeEvolvingArrayType(evolvedType);

```

Comment:
What does returning autoArrayType do here? I thought that it was used *within* control flow.

Reply:
Not exclusively! `getFinalArrayType` called by `finalizeEvolvingArrayType` returns it, too!

Reply:
(Which you'll note is called on the same line in the other ternary branch! We were inconsistent here!)